### ISSUE TITLE
is_a? should warn/error out if it could never possibly return true

### ISSUE THREAD

**Issue by kirbyfan64** [manastech/crystal#989]
This compiles but makes no sense:

```crystal
puts 5.is_a? String
```

This has also bitten me as a bug before. IMO, it should be an error. It's usually a side effect of a code change and is pretty much never intended.

**Comment by vendethiel** [comment]
:+1: I deeply agree.

**Comment by asterite** [comment]
```
- Is 5 a String?
- No
```

I don't see why you think this question makes no sense.

On the other hand, `5 as String` does give a compile error.

Can you show some code that is bug prone due to this behaviour?

**Comment by kirbyfan64** [comment]
@asterite I don't have it anymore, as I fixed it before I uploaded the code online. However, it went kind of like this:

I had `type_eq` function. The first time I used, I accidentally gave it the wrong arguments. That was fine...but it had instances of `is_a?`. I believe I had given it an `Item`, but I meant to give it a `Type`. All the `is_a?`'s failed, and my compiler gave an error kind of like this:

```
error: incompatible types Int and Int in binary operation
```

It took quite a few `puts` calls to figure out that was the issue.

**Comment by vyp** [comment]
> I don't have it anymore, as I fixed it before I uploaded the code online.

Off topic and you probably already know this but if the code is expected to be public you can and should use git offline. :)

Even if it is not expected to become public, git can still push and pull from local directories. Or you could privately self-host, or there are sites which offer free private git hosting (bitbucket, gitlab..). Of course, maybe you just don't want to bother with vcs in early stages, which I completely understand. Just saying in case it's helpful.

**Comment by asterite** [comment]
I see.

Well, the problem is this. Take this code:

```crystal
def foo(x)
  if x.is_a?(String)
    puts 1
  else
    puts 2
  end
end

foo "Hello"
foo 1
```

The above prints 1 and 2, of course. Is this valid code? Should it give an error or a warning?

Well, the way the compiler works is it instantiates "foo" with x begin a String for the first call, and it instantiates "foo" with x being an Int32 in the second call. In the second call "x" is clearly an Int32, so `x.is_a?(String)` is false, but I don't think that code should give an error, not even a warning.

So maybe you can suggest "Well, if the compiler detects multiple calls to `foo` with different types, don't give an error". But then if you have a program with just `foo 1`, it will give an error, but again, it shouldn't.

In short, I don't think there's an easy solution to this.

@vyp Did you intend to comment another issue, maybe?

**Comment by vyp** [comment]
@asterite Sorry, I know it was off-topic. My opinion on this is I'm more leaning towards what you just said, hence why I didn't say anything about the actual issue.

**Comment by asterite** [comment]
@vyp Oh, sorry, I started reading and the forgot that you said "off topic".

**Comment by kirbyfan64** [comment]
@asterite Ah, I see.

I'm confused though. I thought that, on the second instantiation, the type of `x` would be `Int|String`, in which case the `is_a?` is still valid?

**Comment by vyp** [comment]
@kirbyfan64 Am I missing something, wouldn't that not become a "function" anymore, if a previous instantiation makes it side effect like that?

**Comment by bcardiff** [comment]
@kirbyfan64 is correct me it I'm wrong, but I think your situation is with unreachable code. If it is unreachable you want an error.  And the `is_a?` is just one kind of it. Like the following example, we can discuss if the compiler should be smarter enough to discover the unreachable code

```
def foo(x : Int)
  if x.is_a?(String)
    puts 1
  end
end

foo(2)
```

But without the type restriction, it is not the same situation. Because potentially it might be reachable.
```
def foo(x)
  if x.is_a?(String)
    puts 1
  end
end

foo(2)
```

Crystal compiler does not analyze methods definition on its own, but rather upon invocation. With concrete types per parameters. And calling different "instantiations" of a def. So expecting to perform that kind of analysis would be currently off-limits. @asterite will probably point me out if I am outdated regarding this.

**Comment by ysbaddaden** [comment]
What about an "unreachable" comand to find out about such dead code then?

**Comment by bcardiff** [comment]
@ysbaddaden, when you have that kind of features you will need to add some sort of pragmas to turn off false positives probably. I would try to avoid that right now in the language. Yet, I would suggest to check if something from llvm can be used for this kind of static analysis.

**Comment by strcmp** [comment]
this issue reminded me of nim's dead code elimination feature, it can be turned on or off with a compiler switch or is implied in release mode. it can also be configured through pragmas, but i don't think that's necessary (compiler switch is fine). sounds like a hard problem but a cool feature :) as long as its configurable to turn on/off  i think it would be useful but even a 'unreachable' command that could provide hints without dropping any code would be a nice tool.

**Comment by asterite** [comment]
I'm not sure if we are talking about the same thing, but the compiler [transforms](https://github.com/manastech/crystal/blob/master/src/compiler/crystal/semantic/after_type_inference_transformer.cr#L545-L576) "always false" or "always true" checks performed with `is_a?` to `false`/`true` literals. Then it [removes](https://github.com/manastech/crystal/blob/master/src/compiler/crystal/semantic/after_type_inference_transformer.cr#L489-L529) dead `then`/`else` branches if an `if`'s condition is `false`/`true`. So these dummy checks are not even performed by the final executable.

If you are talking about detecting unreachable code, it's not simple to do (or doesn't make much sense) in a language like Crystal. Basically, any method that you don't use, the compiler will tell you that it's unreachable. But that doesn't mean that's an error, not even a warning.

**Comment by bcardiff** [comment]
And that's the reason I couldn't do `while true ... ` simple samples... :+1: 

I agree that more complicated codes are hard to deal. The death code might be on some, but not all, instantiations of a def.

**Comment by waterlink** [comment]
@asterite Methods and classes can be defined as "public API" explicitly. In that case it is safe to fail with compile error if some method is not used and it is not "public API". This error can be quite annoying though.

**Comment by asterite** [comment]
I'll close this. Doing `5.is_a?(Int32)` is fine, and it's also fine if it's always false for a particular program, because it might not be in the general case.
